Add unique exit codes

milljm opened the pull request:
The TestHarness will now exit with specific non-zero exit codes based
on what failed.

Closes #13223

Commit:
Add unique exit codes

The TestHarness will now exit with specific non-zero exit codes based
on what failed.

Closes #13223

milljm commented:
I need to test this with the QueueManager class. Please do not merge (like this would actually pass)

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -207,7 +207,8 @@ def __init__(self, argv, moose_dir, app_name=None):
     0x0  - Success
-    0x80 - TestHarness error
+    0x1  - Recoverable errors
+    0x80 - Unrecoverable TestHarness errors
```
This comment isn't quite accurate. It should be range based.
0x01 - 0x7F - Recoverable errors
0x80 - 0xFF - Unrecoverable TestHarness errors

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -466,9 +463,10 @@ def normalizeStatus(self, job):
-        if job.isFail():
+        if job.isTimeout():
+            tester.setStatus(tester.no_status, 'FAILED (TIMEOUT)')
+        elif job.isFail():
```
This is probably ok for now, but the Job should actually be handling this logic, not the TestHarness.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -484,6 +482,7 @@ def handleJobStatus(self, job):
             # Print results and perform any desired post job processing
             if job.isFinished():
                 tester = self.normalizeStatus(job)
+                self.error_code = self.error_code | job.getStatusCode() | tester.getStatusCode()
```
Again, this should be moved into the Job. There's too much "leaky encapsulation" going on here. Don't make the TestHarness dig or figure this kind of stuff out. The Job should do it and then the TestHarness can just make a simple query to get the desired information back.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -498,11 +497,11 @@ def handleJobStatus(self, job):
                 elif tester.isPass():
                     self.num_passed += 1
-                elif tester.isFail():
+                elif tester.isFail() or job.isTimeout():
```
This logic is probably ok here. Up to you.

permcody commented on python/TestHarness/TestHarness.py:
```diff
@@ -539,11 +538,9 @@ def cleanup(self):
         if len(self.parse_errors) > 0:
             fatal_error += ', <r>FATAL PARSER ERROR</r>'
-            self.error_code = 1
+            self.error_code = self.error_code | 0x80
```
Doesn't have to be 0x80. It can be anything higher than that.

permcody commented on python/TestHarness/schedulers/Job.py:
```diff
@@ -241,6 +242,8 @@ def getStatus(self):
         return self.status.getStatusMessage()
+    def getStatusCode(self):
+        return self.status.getStatusCode()
     def setStatus(self, status, message=''):
```
Who calls this method? It should probably be documented. This method shouldn't be used by most objects (including the TestHarness). Maybe the Scheduler uses it?

permcody commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -302,7 +302,7 @@ def handleTimeoutJob(self, job, j_lock):
         with j_lock:
             if job.isRunning():
-                job.setStatus(job.crash, 'TIMEOUT')
+                job.setStatus(job.timeout, 'TIMEOUT')
```
Ah! Answered my own question, the Scheduler calls the set method.

permcody commented on python/TestHarness/tests/test_Deleted.py:
```diff
@@ -21,6 +21,9 @@ def testDeleted(self):
         self.assertRegexpMatches(e.output, r'test_harness\.deleted.*? \[TEST DELETED TEST\] FAILED \(DELETED\)')
 
+        # Verify return code is DELETED related (0x83)
+        self.assertIs(131, e.returncode)
```
Why not pass 0x83 in here instead of using decimal?

permcody commented on python/TestHarness/tests/test_Diff.py:
```diff
@@ -23,3 +23,6 @@ def testDiffs(self):
         self.assertRegexpMatches(e.output, r'test_harness\.csvdiff.*?Running csvdiff')
+
+        # Verify return code is DIFF related (0x81)
+        self.assertIs(129, e.returncode)
```
Use Hex

permcody commented on python/TestHarness/tests/test_MissingGold.py:
```diff
@@ -21,3 +21,6 @@ def testMissingGold(self):
         self.assertRegexpMatches(e.output, 'test_harness\.csvdiff.*?FAILED \(MISSING GOLD FILE\)')
+
+        # Verify return code is a general failure related (0x80)
+        self.assertIs(128, e.returncode)
```
Use Hex

permcody reviewed (commented):
This isn't bad. At least it's not any worse than some of the existing design issues we already have. We should work to move more of the Tester handling logic into the Job class. A better design would insulate the TestHarness from the internals.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/320828/) on b0ce231 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13243/site)

This comment will be updated on new commits.

Commit:
Simplify status passing

Move the status handling code into job in an attempt to simplify the process
when the TestHarness asks for the current status. To do this, we need a new
way to store/retrieve a status (without modifying the behavior for either the
Tester or Scheduler status-handling).

Commit:
Job class to control status comparisons

Create a unified StatusSystem class, which will provide status objects containing
a status, the color of said status, and a proper exit code.

Shift the responsibility of status comparisons to the job class. The TestHarness
(and anything else requiring status formatting) needs only call getJoinStatus()
for all its output status printing needs.

QueueManager creates a proper status via the StatusSystem (instead of creating
it own -might-be-compatible- namedtuple)

permcody commented:
Well it passed!

milljm commented:
I was fairly thorough spending a week on it!

permcody commented on python/TestHarness/schedulers/Job.py:
```diff
@@ -236,28 +253,76 @@ def getTiming(self):
+
+    def getJointStatus(self):
+        """
+        Return the most accurate message possible, based on the statuses of both the Tester and Job.
```
This is much better.

permcody commented:
Much better - now go ahead and update the recipes so that we can try this out!
